libyul/backends/evm/StackLayoutGenerator.h

```diff
@@ -0,0 +1,95 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Stack layout generator for Yul to EVM code generation.
+ */
+
+#pragma once
+
+#include <libyul/backends/evm/ControlFlowGraph.h>
+
+#include <map>
+
+namespace solidity::yul
+{
+
+struct StackLayout
+{
+	struct BlockInfo
+	{
+		/// Complete stack layout that is required for entering a block.
+		Stack entryLayout;
+		/// The resulting stack layout after executing the block.
+		Stack exitLayout;
+	};
+	std::map<CFG::BasicBlock const*, BlockInfo> blockInfos;
+	/// For each operation the complete stack layout that:
+	/// - has the slots required for the operation at the stack top.
+	/// - will have the operation result in a layout that makes it easy to achieve the next desired layout.
+	std::map<CFG::Operation const*, Stack> operationEntryLayout;
+};
+
+class StackLayoutGenerator
+{
+public:
+	static StackLayout run(CFG const& _cfg);
+
+private:
+	StackLayoutGenerator(StackLayout& _context);
+
+	/// @returns the optimal entry stack layout, s.t. @a _operation can be applied to it and
+	/// the result can be transformed to @a _exitStack with minimal stack shuffling.
+	Stack propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation);
+
+	/// @returns the desired stack layout at the entry of @a _block, assuming the layout after
+	/// executing the block should be @a _exitStack.
+	Stack propagateStackThroughBlock(Stack _exitStack, CFG::BasicBlock const& _block);
+
+	/// Main algorithm walking the graph from entry to exit and propagating back the stack layouts to the entries.
+	/// Iteratively reruns itself along backwards jumps until the layout is stabilized.
+	void processEntryPoint(CFG::BasicBlock const& _entry);
+
+	std::optional<Stack> getExitLayoutOrStageDependencies(
+		CFG::BasicBlock const& _block,
+		std::set<CFG::BasicBlock const*> const& _blocksWithExitLayouts,
```

Comment:
It's a bit confusing that this is called `visited` in the caller.

Reply:
I'm actually not sure anymore if making this an independent function is that useful...

Reply:
Yeah... I'm also not so sure... I'll keep it separate for now, but I'm definitely fine with pulling it back in - not sure there's a way around passing both ``_visited`` and ``_toVisit``...